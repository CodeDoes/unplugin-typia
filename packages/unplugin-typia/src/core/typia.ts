import ts from 'typescript';
import { readTSConfig } from 'pkg-types';
import MagicString from 'magic-string';
import type { UnpluginBuildContext, UnpluginContext } from 'unplugin';
import { transform as typiaTransform } from 'typia/lib/transform.js';

import type { ResolvedOptions } from './options.ts';

/** create a printer */
const printer = ts.createPrinter();

/** cache compilerOptions */
let compilerOptions: ts.CompilerOptions | undefined;

/** cache source files */
const sourceCache = new Map<string, ts.SourceFile>();

/**
 * Transform a TypeScript file with Typia.
 *
 * @param id - The file path.
 * @param source - The source code.
 * @param unpluginContext - The unplugin context.
 * @param options - The resolved options.
 * @returns The transformed code and source map.
 */
export async function transformTypia(
	id: string,
	source: string,
	/**
	 * **Use with caution.**
	 *
	 * This is an experimental feature and may be changed at any time.
	 */
	unpluginContext: UnpluginBuildContext & UnpluginContext,
	options: ResolvedOptions,
): Promise<{ code: string; map: any } | undefined> {
	/** Whether to enable cache */
	const cacheEnable = options.cache.enable;

	/** parse tsconfig compilerOptions */
	compilerOptions = await getTsCompilerOption(cacheEnable);

	const { program, tsSource } = await getProgramAndSource(id, source, compilerOptions, cacheEnable);

	const {
		diagnostics,
		transformed,
		file,
	} = transform(id, program, tsSource, options.typia);

	warnDiagnostic(diagnostics, transformed, unpluginContext);

	const generatedSource = printer.printFile(file);
	const magic = new MagicString(generatedSource);

	return {
		code: magic.toString(),
		map: magic.generateMap({
			source: id,
			file: `${id}.map`,
		}),
	};
}

/**
 * Read tsconfig.json and get compilerOptions.
 * @param cacheEnable - Whether to enable cache. @default true
 */
async function getTsCompilerOption(cacheEnable = true): Promise<ts.CompilerOptions> {
	const parseTsComilerOptions = async () => (({ ...(await readTSConfig())?.compilerOptions, moduleResolution: undefined }));

	/** parse tsconfig compilerOptions */
	if (cacheEnable) {
		compilerOptions ??= await parseTsComilerOptions();
	}
	else {
		compilerOptions = await parseTsComilerOptions();
	}

	if (compilerOptions == null) {
		throw new Error('No compilerOptions found in tsconfig.json');
	}
	return compilerOptions;
}

/**
 * Get program and source.
 *
 * @param id - The file path.
 * @param source - The source code.
 * @param compilerOptions - The compiler options.
 * @param cacheEnable - Whether to enable cache. @default true
 * @returns The program and source.
 */
async function getProgramAndSource(
	id: string,
	source: string,
	compilerOptions: ts.CompilerOptions,
	cacheEnable = true,
): Promise<{ program: ts.Program; tsSource: ts.SourceFile }> {
	const tsSource = ts.createSourceFile(
		id,
		source,
		ts.ScriptTarget.ES2020,
	);
	const host = ts.createCompilerHost(compilerOptions);

	host.getSourceFile = (fileName, languageVersion) => {
		if (fileName === id) {
			return tsSource;
		}

		if (cacheEnable) {
			const cache = sourceCache.get(fileName);
			if (cache != null) {
				return cache;
			}
		}

		const source = ts.sys.readFile(fileName);
		if (source == null) {
			return undefined;
		}
		const result = ts.createSourceFile(fileName, source, languageVersion);

		if (cacheEnable) {
			sourceCache.set(fileName, result);
		}

		return result;
	};
	const program = ts.createProgram([id], compilerOptions, host);

	return { program, tsSource };
}

/**
 * Transform a TypeScript file with Typia.
 *
 * @param id - The file path.
 * @param program - The program.
 * @param tsSource - The source file.
 * @param typiaOptions - The Typia options.
 * @returns The transformed code and source map.
 */
function transform(
	id: string,
	program: ts.Program,
	tsSource: ts.SourceFile,
	typiaOptions?: ResolvedOptions['typia'],
): {
	/** The diagnostics */
		diagnostics: ts.Diagnostic[];
		/** The transformed source files */
		transformed: ts.SourceFile[];
		/** The transformed source file we need */
		file: ts.SourceFile;
	} {
	const diagnostics: ts.Diagnostic[] = [];

	/** transform with Typia */
	const typiaTransformed = typiaTransform(program, typiaOptions, {
		addDiagnostic(diag) {
			return diagnostics.push(diag);
		},
	});

	/** transform with TypeScript */
	const transformationResult = ts.transform(
		tsSource,
		[typiaTransformed],
		{
			...program.getCompilerOptions(),
			sourceMap: true,
			inlineSources: true,
		},
	);

	const file = transformationResult.transformed.find(t => t.fileName === id);

	if (file == null) {
		throw new Error('No file found');
	}

	/** dispose transformation result */
	transformationResult.dispose();

	const { transformed } = transformationResult;

	return { diagnostics, transformed, file };
}

/** Warn diagnostics */
function warnDiagnostic(
	diagnostics: ts.Diagnostic[],
	transformed: ts.SourceFile[],
	unpluginContext: UnpluginBuildContext & UnpluginContext,
) {
	for (const diagnostic of diagnostics) {
		unpluginContext.warn(
			transformed.map(e => e.fileName).join(','),
		);
		unpluginContext.warn(JSON.stringify(diagnostic.messageText));
	}
}
