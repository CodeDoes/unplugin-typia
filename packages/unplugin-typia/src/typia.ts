import * as ts from 'typescript';
import * as U from '@core/unknownutil';
import MagicString from 'magic-string';
import { transform } from 'typia/lib/transform.js';

const printer = ts.createPrinter();

export function transformTypia(
	id: string,
	service: ts.LanguageService,
): { code: string; map: any } | undefined {
	const program = service.getProgram();

	const tsSource = program?.getSourceFile(id);

	if (U.isNullish(tsSource)) {
		throw new Error('No source found');
	}

	if (U.isNullish(program)) {
		throw new Error('No program found');
	}

	const diagnostics: ts.Diagnostic[] = [];

	const typiaTransformed = transform(program, {}, {
		addDiagnostic(diag) {
			return diagnostics.push(diag);
		},
	});
	const transformed = ts.transform(
		tsSource,
		[typiaTransformed],
		{
			...program.getCompilerOptions(),
			sourceMap: true,
			inlineSources: true,
		},
	);

	const file = transformed.transformed.find(t => t.fileName === id);

	if (U.isNullish(file)) {
		throw new Error('No file found');
	}

	const generatedSource = printer.printFile(file);
	for (const diagnostic of diagnostics) {
		console.warn(transformed.transformed.map(e => e.fileName).join(','));
		console.warn(JSON.stringify(diagnostic.messageText));
	}

	const magic = new MagicString(generatedSource);

	transformed.dispose();

	return {
		code: magic.toString(),
		map: magic.generateMap({
			source: id,
			file: `${id}.map`,
		}),
	};
}
