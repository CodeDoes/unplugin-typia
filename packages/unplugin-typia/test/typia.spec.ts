import fs from 'node:fs/promises';
import { expect, it } from 'vitest';
import { x } from 'tinyexec';

import type { UnpluginBuildContext, UnpluginContext } from 'unplugin';
import { join } from 'pathe';
import { transformTypia } from '../src/core/typia.js';
import { resolveOptions } from '../src/api.js';
import type { Data, ID, Source } from '../src/core/types.js';

class DummyContext {
	warn(args: unknown) {
		console.warn(args);
	}
};

const ctx = new DummyContext() as UnpluginContext & UnpluginBuildContext;

async function test(_id: string): Promise<Data> {
	const id = join(import.meta.dirname, './fixtures/', _id) as ID;
	const code = await fs.readFile(id, 'utf-8') as Source;
	const transformed = await transformTypia(id, code, ctx, resolveOptions({ cache: false }));
	return transformed;
}

function getSnapshotPath(id: string): string {
	return join(import.meta.dirname, './snapshot/', id);
}

it('transform is', async () => {
	const transformed = await test('is.ts');
	const snapshot = getSnapshotPath('is.ts');
	await expect(transformed).toMatchFileSnapshot(snapshot);
	await x('bun', [snapshot]);
});

it('transform validate', async () => {
	const transformed = await test('validate.ts');
	const snapshot = getSnapshotPath('validate.ts');
	await expect(transformed).toMatchFileSnapshot(snapshot);
	await x('bun', [snapshot]);
});

it('transform random', async () => {
	const transformed = await test('random.ts');
	const snapshot = getSnapshotPath('random.ts');
	await expect(transformed).toMatchFileSnapshot(snapshot);
	await x('bun', [snapshot]);
});
