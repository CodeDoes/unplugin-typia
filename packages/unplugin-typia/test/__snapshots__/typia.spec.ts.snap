// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Transform validate.ts 1`] = `
"import typia from 'typia';
import type { IMember } from './type.js';
const validate = (input: any): typia.IValidation<IMember> => {
    const errors = [] as any[];
    const __is = (input: any): input is IMember => {
        return "object" === typeof input && null !== input && ("string" === typeof (input as any).email && /^[a-z0-9!#$%&'*+/=?^_\`{|}~-]+(?:/.[a-z0-9!#$%&'*+/=?^_\`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?$/i.test((input as any).email) && ("string" === typeof (input as any).id && /^(?:urn:uuid:)?[0-9a-f]{8}-(?:[0-9a-f]{4}-){3}[0-9a-f]{12}$/i.test((input as any).id)) && ("number" === typeof (input as any).age && (Math.floor((input as any).age) === (input as any).age && 0 <= (input as any).age && (input as any).age <= 4294967295 && 19 < (input as any).age && (input as any).age <= 100)));
    };
    if (false === __is(input)) {
        const $report = (typia.createValidate as any).report(errors);
        ((input: any, _path: string, _exceptionable: boolean = true): input is IMember => {
            const $vo0 = (input: any, _path: string, _exceptionable: boolean = true): boolean => ["string" === typeof input.email && (/^[a-z0-9!#$%&'*+/=?^_\`{|}~-]+(?:/.[a-z0-9!#$%&'*+/=?^_\`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?$/i.test(input.email) || $report(_exceptionable, {
                    path: _path + ".email",
                    expected: "string & Format</"email/">",
                    value: input.email
                })) || $report(_exceptionable, {
                    path: _path + ".email",
                    expected: "(string & Format</"email/">)",
                    value: input.email
                }), "string" === typeof input.id && (/^(?:urn:uuid:)?[0-9a-f]{8}-(?:[0-9a-f]{4}-){3}[0-9a-f]{12}$/i.test(input.id) || $report(_exceptionable, {
                    path: _path + ".id",
                    expected: "string & Format</"uuid/">",
                    value: input.id
                })) || $report(_exceptionable, {
                    path: _path + ".id",
                    expected: "(string & Format</"uuid/">)",
                    value: input.id
                }), "number" === typeof input.age && (Math.floor(input.age) === input.age && 0 <= input.age && input.age <= 4294967295 || $report(_exceptionable, {
                    path: _path + ".age",
                    expected: "number & Type</"uint32/">",
                    value: input.age
                })) && (19 < input.age || $report(_exceptionable, {
                    path: _path + ".age",
                    expected: "number & ExclusiveMinimum<19>",
                    value: input.age
                })) && (input.age <= 100 || $report(_exceptionable, {
                    path: _path + ".age",
                    expected: "number & Maximum<100>",
                    value: input.age
                })) || $report(_exceptionable, {
                    path: _path + ".age",
                    expected: "(number & Type</"uint32/"> & ExclusiveMinimum<19> & Maximum<100>)",
                    value: input.age
                })].every((flag: boolean) => flag);
            return ("object" === typeof input && null !== input || $report(true, {
                path: _path + "",
                expected: "IMember",
                value: input
            })) && $vo0(input, _path + "", true) || $report(true, {
                path: _path + "",
                expected: "IMember",
                value: input
            });
        })(input, "$input", true);
    }
    const success = 0 === errors.length;
    return {
        success,
        errors,
        data: success ? input : undefined
    } as any;
};
validate({});"
`;

exports[`Transform type.d.ts 1`] = `
"import type { tags } from 'typia';
export type IMember = {
    email: string & tags.Format<'email'>;
    id: string & tags.Format<'uuid'>;
    age: number & tags.Type<'uint32'> & tags.ExclusiveMinimum<19> & tags.Maximum<100>;
};"
`;

exports[`Transform random.ts 1`] = `
"import typia from 'typia';
import type { IMember } from './type.js';
const random = (generator?: Partial<typia.IRandomGenerator>): import("typia").Resolved<IMember> => {
    const $generator = (typia.createRandom as any).generator;
    const $ro0 = (_recursive: boolean = false, _depth: number = 0): any => ({
        email: (generator?.customs ?? $generator.customs)?.string?.([
            {
                name: "Format</"email/">",
                kind: "format",
                value: "email"
            }
        ]) ?? (generator?.email ?? $generator.email)(),
        id: (generator?.customs ?? $generator.customs)?.string?.([
            {
                name: "Format</"uuid/">",
                kind: "format",
                value: "uuid"
            }
        ]) ?? (generator?.uuid ?? $generator.uuid)(),
        age: (generator?.customs ?? $generator.customs)?.number?.([
            {
                name: "Type</"uint32/">",
                kind: "type",
                value: "uint32"
            },
            {
                name: "ExclusiveMinimum<19>",
                kind: "exclusiveMinimum",
                value: 19
            },
            {
                name: "Maximum<100>",
                kind: "maximum",
                value: 100
            }
        ]) ?? (generator?.integer ?? $generator.integer)(19, 100)
    });
    return $ro0();
};
random();"
`;

exports[`Transform is.ts 1`] = `
"import typia from 'typia';
import type { IMember } from './type.js';
const is = (input: any): input is IMember => {
    return "object" === typeof input && null !== input && ("string" === typeof (input as any).email && /^[a-z0-9!#$%&'*+/=?^_\`{|}~-]+(?:/.[a-z0-9!#$%&'*+/=?^_\`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?$/i.test((input as any).email) && ("string" === typeof (input as any).id && /^(?:urn:uuid:)?[0-9a-f]{8}-(?:[0-9a-f]{4}-){3}[0-9a-f]{12}$/i.test((input as any).id)) && ("number" === typeof (input as any).age && (Math.floor((input as any).age) === (input as any).age && 0 <= (input as any).age && (input as any).age <= 4294967295 && 19 < (input as any).age && (input as any).age <= 100)));
};
is({});"
`;
